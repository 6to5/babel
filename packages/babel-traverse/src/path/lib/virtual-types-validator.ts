import type NodePath from "../index";
import {
  isBinding,
  isBlockScoped as nodeIsBlockScoped,
  isExportDeclaration,
  isExpression as nodeIsExpression,
  isFlow as nodeIsFlow,
  isForStatement,
  isForXStatement,
  isIdentifier,
  isImportDeclaration,
  isImportSpecifier,
  isJSXIdentifier,
  isJSXMemberExpression,
  isMemberExpression,
  isReferenced as nodeIsReferenced,
  isScope as nodeIsScope,
  isStatement as nodeIsStatement,
  isVar as nodeIsVar,
  isVariableDeclaration,
  react,
} from "@babel/types";
import type * as t from "@babel/types";
const { isCompatTag } = react;
import type { VirtualTypeAliases } from "./virtual-types";

export interface VirtualTypeNodePathValidators {
  isBindingIdentifier(
    opts?: object,
  ): this is NodePath<VirtualTypeAliases["BindingIdentifier"]>;
  isBlockScoped(opts?: object): boolean;
  isExistentialTypeParam(
    opts?: object,
  ): this is NodePath<VirtualTypeAliases["ExistentialTypeParam"]>;
  isExpression(opts?: object): this is NodePath<t.Expression>;
  isFlow(opts?: object): this is NodePath<t.Flow>;
  isForAwaitStatement(
    opts?: object,
  ): this is NodePath<VirtualTypeAliases["ForAwaitStatement"]>;
  isGenerated(opts?: object): boolean;
  isNumericLiteralTypeAnnotation(
    opts?: object,
  ): this is NodePath<VirtualTypeAliases["NumericLiteralTypeAnnotation"]>;
  isPure(opts?: object): boolean;
  isReferenced(opts?: object): boolean;
  isReferencedIdentifier(
    opts?: object,
  ): this is NodePath<VirtualTypeAliases["ReferencedIdentifier"]>;
  isReferencedMemberExpression(
    opts?: object,
  ): this is NodePath<VirtualTypeAliases["ReferencedMemberExpression"]>;
  isRestProperty(opts?: object): this is NodePath<t.RestProperty>;
  isScope(opts?: object): this is NodePath<VirtualTypeAliases["Scope"]>;
  isSpreadProperty(opts?: object): this is NodePath<t.SpreadProperty>;
  isStatement(opts?: object): this is NodePath<t.Statement>;
  isUser(opts?: object): boolean;
  isVar(opts?: object): this is NodePath<VirtualTypeAliases["Var"]>;
}

export function isReferencedIdentifier(this: NodePath, opts?: any): boolean {
  const { node, parent } = this;
  if (!isIdentifier(node, opts) && !isJSXMemberExpression(parent, opts)) {
    if (isJSXIdentifier(node, opts)) {
      if (isCompatTag(node.name)) return false;
    } else {
      // not a JSXIdentifier or an Identifier
      return false;
    }
  }

  // check if node is referenced
  return nodeIsReferenced(node, parent, this.parentPath.parent);
}

export function isReferencedMemberExpression(this: NodePath): boolean {
  const { node, parent } = this;
  return isMemberExpression(node) && nodeIsReferenced(node, parent);
}

export function isBindingIdentifier(this: NodePath): boolean {
  const { node, parent } = this;
  const grandparent = this.parentPath.parent;
  return isIdentifier(node) && isBinding(node, parent, grandparent);
}

export function isStatement(this: NodePath): boolean {
  const { node, parent } = this;
  if (nodeIsStatement(node)) {
    if (isVariableDeclaration(node)) {
      if (isForXStatement(parent, { left: node })) return false;
      if (isForStatement(parent, { init: node })) return false;
    }

    return true;
  } else {
    return false;
  }
}

export function isExpression(this: NodePath): boolean {
  if (this.isIdentifier()) {
    return this.isReferencedIdentifier();
  } else {
    return nodeIsExpression(this.node);
  }
}

export function isScope(this: NodePath): boolean {
  return nodeIsScope(this.node, this.parent);
}

export function isReferenced(this: NodePath): boolean {
  return nodeIsReferenced(this.node, this.parent);
}

export function isBlockScoped(this: NodePath): boolean {
  return nodeIsBlockScoped(this.node, this.parent);
}

export function isVar(this: NodePath): boolean {
  return nodeIsVar(this.node);
}

export function isUser(this: NodePath): boolean {
  return this.node && !!this.node.loc;
}

export function isGenerated(this: NodePath): boolean {
  return !this.isUser();
}

export function isPure(this: NodePath, opts?): boolean {
  return this.scope.isPure(this.node, opts);
}

export function isFlow(this: NodePath): boolean {
  const { node } = this;
  if (nodeIsFlow(node)) {
    return true;
  } else if (isImportDeclaration(node)) {
    return node.importKind === "type" || node.importKind === "typeof";
  } else if (isExportDeclaration(node)) {
    return node.exportKind === "type";
  } else if (isImportSpecifier(node)) {
    return node.importKind === "type" || node.importKind === "typeof";
  } else {
    return false;
  }
}

// TODO: 7.0 Backwards Compat
// todo: we should check RestProperty first
export function isRestProperty(this: NodePath): boolean {
  return this.parentPath && this.parentPath.isObjectPattern();
}

// todo: we should check RestProperty first
export function isSpreadProperty(this: NodePath): boolean {
  return this.parentPath && this.parentPath.isObjectExpression();
}

// todo: we should check isForOfStatement first
export function isForAwaitStatement(this: NodePath<t.ForOfStatement>): boolean {
  return this.node.await === true;
}
