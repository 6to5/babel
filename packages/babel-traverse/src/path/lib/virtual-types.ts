export interface VirtualTypeAliases {
  BindingIdentifier: t.Identifier;
  BlockScoped: t.Node;
  ExistentialTypeParam: t.ExistsTypeAnnotation;
  Expression: t.Expression;
  Flow: t.Flow | t.ImportDeclaration | t.ExportDeclaration | t.ImportSpecifier;
  ForAwaitStatement: t.ForOfStatement;
  Generated: t.Node;
  NumericLiteralTypeAnnotation: t.NumberLiteralTypeAnnotation;
  Pure: t.Node;
  Referenced: t.Node;
  ReferencedIdentifier: t.Identifier | t.JSXIdentifier;
  ReferencedMemberExpression: t.MemberExpression;
  RestProperty: t.RestElement;
  Scope: t.Scopable | t.Pattern;
  SpreadProperty: t.RestElement;
  Statement: t.Statement;
  User: t.Node;
  Var: t.VariableDeclaration;
}

type NodeTypes = t.Node["type"] | t.Comment["type"] | keyof t.Aliases;

export type Wrapper = {
  types?: NodeTypes[];
  checkPath?(path: NodePath): boolean;
};

export const ReferencedIdentifier: Wrapper = {
  types: ["Identifier", "JSXIdentifier"],
  checkPath: path => path.isReferencedIdentifier(),
};

export const ReferencedMemberExpression: Wrapper = {
  types: ["MemberExpression"],
  checkPath: path => path.isReferencedMemberExpression(),
};

export const BindingIdentifier: Wrapper = {
  types: ["Identifier"],
  checkPath: path => path.isBindingIdentifier(),
};

export const Statement: Wrapper = {
  types: ["Statement"],
  checkPath: path => path.isStatement(),
};

export const Expression: Wrapper = {
  types: ["Expression"],
  checkPath: path => path.isExpression(),
};

export const Scope: Wrapper = {
  // When pattern is inside the function params, it is a scope
  types: ["Scopable", "Pattern"],
  checkPath: path => path.isScope(),
};

export const Referenced: Wrapper = {
  checkPath: path => path.isReferenced(),
};

export const BlockScoped: Wrapper = {
  checkPath: path => path.isBlockScoped(),
};

export const Var: Wrapper = {
  types: ["VariableDeclaration"],
  checkPath: path => path.isVar(),
};

export const User: Wrapper = {
  checkPath: path => path.isUser(),
};

export const Generated: Wrapper = {
  checkPath: path => path.isGenerated(),
};

export const Pure: Wrapper = {
  checkPath: path => path.isPure(),
};

export const Flow: Wrapper = {
  types: ["Flow", "ImportDeclaration", "ExportDeclaration", "ImportSpecifier"],
  checkPath: path => path.isFlow(),
};

// TODO: 7.0 Backwards Compat
export const RestProperty: Wrapper = {
  types: ["RestElement"],
  checkPath: path => path.isRestProperty(),
};

export const SpreadProperty: Wrapper = {
  types: ["RestElement"],
  checkPath: path => path.isSpreadProperty(),
};

export const ExistentialTypeParam: Wrapper = {
  types: ["ExistsTypeAnnotation"],
};

export const NumericLiteralTypeAnnotation: Wrapper = {
  types: ["NumberLiteralTypeAnnotation"],
};

export const ForAwaitStatement: Wrapper = {
  types: ["ForOfStatement"],
  checkPath: path => path.isForAwaitStatement(),
};
