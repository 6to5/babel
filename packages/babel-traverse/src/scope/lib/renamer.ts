import Binding from "../binding";
import splitExportDeclaration from "@babel/helper-split-export-declaration";
import * as t from "@babel/types";
import type { NodePath, Visitor } from "../..";
import { requeueComputedKeyAndDecorators } from "@babel/helper-environment-visitor";

const renameVisitor: Visitor<Renamer> = {
  ReferencedIdentifier({ node }, state) {
    if (node.name === state.oldName) {
      node.name = state.newName;
    }
  },

  Scope(path, state) {
    if (
      !path.scope.bindingIdentifierEquals(
        state.oldName,
        state.binding.identifier,
      )
    ) {
      path.skip();
      if (path.isMethod()) {
        requeueComputedKeyAndDecorators(path);
      }
    }
  },

  "AssignmentExpression|Declaration|VariableDeclarator"(
    path: NodePath<t.AssignmentPattern | t.Declaration | t.VariableDeclarator>,
    state,
  ) {
    if (path.isVariableDeclaration()) return;
    const ids = path.getOuterBindingIdentifiers();

    for (const name in ids) {
      if (name === state.oldName) ids[name].name = state.newName;
    }
  },
};

export default class Renamer {
  constructor(binding: Binding, oldName: string, newName: string) {
    this.newName = newName;
    this.oldName = oldName;
    this.binding = binding;
  }

  declare oldName: string;
  declare newName: string;
  declare binding: Binding;

  maybeConvertFromExportDeclaration(parentDeclar: NodePath) {
    const maybeExportDeclar = parentDeclar.parentPath;

    if (!maybeExportDeclar.isExportDeclaration()) {
      return;
    }

    if (maybeExportDeclar.isExportDefaultDeclaration()) {
      const { declaration } = maybeExportDeclar.node;
      if (t.isDeclaration(declaration) && !declaration.id) {
        return;
      }
    }

    if (maybeExportDeclar.isExportAllDeclaration()) {
      return;
    }

    splitExportDeclaration(
      maybeExportDeclar as NodePath<
        Exclude<t.ExportDeclaration, t.ExportAllDeclaration>
      >,
    );
  }

  maybeConvertFromClassFunctionDeclaration() {
    return; // TODO
  }

  maybeConvertFromClassFunctionExpression() {
    return; // TODO
  }

  rename(block?: t.Pattern | t.Scopable) {
    const { binding, oldName, newName } = this;
    const { scope, path } = binding;

    const parentDeclar = path.find(
      path =>
        path.isDeclaration() ||
        path.isFunctionExpression() ||
        path.isClassExpression(),
    );
    if (parentDeclar) {
      const bindingIds = parentDeclar.getOuterBindingIdentifiers();
      if (bindingIds[oldName] === binding.identifier) {
        // When we are renaming an exported identifier, we need to ensure that
        // the exported binding keeps the old name.
        this.maybeConvertFromExportDeclaration(parentDeclar);
      }
    }

    const blockToTraverse = block || scope.block;
    if (blockToTraverse?.type === "SwitchStatement") {
      // discriminant is not part of current scope, should be skipped.
      blockToTraverse.cases.forEach(c => {
        scope.traverse(c, renameVisitor, this);
      });
    } else {
      scope.traverse(blockToTraverse, renameVisitor, this);
    }

    if (!block) {
      scope.removeOwnBinding(oldName);
      scope.bindings[newName] = binding;
      this.binding.identifier.name = newName;
    }

    if (parentDeclar) {
      this.maybeConvertFromClassFunctionDeclaration();
      this.maybeConvertFromClassFunctionExpression();
    }
  }
}
