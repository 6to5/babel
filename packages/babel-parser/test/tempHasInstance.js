import { parse } from "../lib";

function getParser(code) {
  return () => parse(code, { sourceType: "module" });
}

const testCode = `
class Range {
    #start
    #end
    constructor(start, end) {
        this.#start = start
        this.#end = end
    }
    test(){
    }
    equals(range) {
        // this.test(range);
        call(range)
        if (!(class.hasInstance(range))) return false
        // return this.#start === range.#start && this.#end === range.#end
    }
}
`;

describe("class function syntax", () => {
  it("should parse", () => {
    expect(getParser(testCode)()).toMatchSnapshot();
  });
});
