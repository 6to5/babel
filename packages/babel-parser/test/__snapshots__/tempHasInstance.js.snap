// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class function syntax should parse 1`] = `
Node {
  "comments": Array [
    Object {
      "end": 189,
      "loc": SourceLocation {
        "end": Position {
          "column": 28,
          "line": 12,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 8,
          "line": 12,
        },
      },
      "start": 169,
      "type": "CommentLine",
      "value": " this.test(range);",
    },
    Object {
      "end": 338,
      "loc": SourceLocation {
        "end": Position {
          "column": 74,
          "line": 15,
        },
        "filename": undefined,
        "identifierName": undefined,
        "start": Position {
          "column": 8,
          "line": 15,
        },
      },
      "start": 272,
      "type": "CommentLine",
      "value": " return this.#start === range.#start && this.#end === range.#end",
    },
  ],
  "end": 348,
  "errors": Array [],
  "loc": SourceLocation {
    "end": Position {
      "column": 0,
      "line": 19,
    },
    "filename": undefined,
    "identifierName": undefined,
    "start": Position {
      "column": 0,
      "line": 1,
    },
  },
  "program": Node {
    "body": Array [
      Node {
        "body": Node {
          "body": Array [
            Node {
              "end": 25,
              "key": Node {
                "end": 25,
                "id": Node {
                  "end": 25,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 10,
                      "line": 3,
                    },
                    "filename": undefined,
                    "identifierName": "start",
                    "start": Position {
                      "column": 5,
                      "line": 3,
                    },
                  },
                  "name": "start",
                  "start": 20,
                  "type": "Identifier",
                },
                "loc": SourceLocation {
                  "end": Position {
                    "column": 10,
                    "line": 3,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 4,
                    "line": 3,
                  },
                },
                "start": 19,
                "type": "PrivateName",
              },
              "loc": SourceLocation {
                "end": Position {
                  "column": 10,
                  "line": 3,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 4,
                  "line": 3,
                },
              },
              "start": 19,
              "static": false,
              "type": "ClassPrivateProperty",
              "value": null,
            },
            Node {
              "end": 34,
              "key": Node {
                "end": 34,
                "id": Node {
                  "end": 34,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 8,
                      "line": 4,
                    },
                    "filename": undefined,
                    "identifierName": "end",
                    "start": Position {
                      "column": 5,
                      "line": 4,
                    },
                  },
                  "name": "end",
                  "start": 31,
                  "type": "Identifier",
                },
                "loc": SourceLocation {
                  "end": Position {
                    "column": 8,
                    "line": 4,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 4,
                    "line": 4,
                  },
                },
                "start": 30,
                "type": "PrivateName",
              },
              "loc": SourceLocation {
                "end": Position {
                  "column": 8,
                  "line": 4,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 4,
                  "line": 4,
                },
              },
              "start": 30,
              "static": false,
              "type": "ClassPrivateProperty",
              "value": null,
            },
            Node {
              "async": false,
              "body": Node {
                "body": Array [
                  Node {
                    "end": 92,
                    "expression": Node {
                      "end": 92,
                      "left": Node {
                        "computed": false,
                        "end": 84,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 19,
                            "line": 6,
                          },
                          "filename": undefined,
                          "identifierName": undefined,
                          "start": Position {
                            "column": 8,
                            "line": 6,
                          },
                        },
                        "object": Node {
                          "end": 77,
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 12,
                              "line": 6,
                            },
                            "filename": undefined,
                            "identifierName": undefined,
                            "start": Position {
                              "column": 8,
                              "line": 6,
                            },
                          },
                          "start": 73,
                          "type": "ThisExpression",
                        },
                        "property": Node {
                          "end": 84,
                          "id": Node {
                            "end": 84,
                            "loc": SourceLocation {
                              "end": Position {
                                "column": 19,
                                "line": 6,
                              },
                              "filename": undefined,
                              "identifierName": "start",
                              "start": Position {
                                "column": 14,
                                "line": 6,
                              },
                            },
                            "name": "start",
                            "start": 79,
                            "type": "Identifier",
                          },
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 19,
                              "line": 6,
                            },
                            "filename": undefined,
                            "identifierName": undefined,
                            "start": Position {
                              "column": 13,
                              "line": 6,
                            },
                          },
                          "start": 78,
                          "type": "PrivateName",
                        },
                        "start": 73,
                        "type": "MemberExpression",
                      },
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 27,
                          "line": 6,
                        },
                        "filename": undefined,
                        "identifierName": undefined,
                        "start": Position {
                          "column": 8,
                          "line": 6,
                        },
                      },
                      "operator": "=",
                      "right": Node {
                        "end": 92,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 27,
                            "line": 6,
                          },
                          "filename": undefined,
                          "identifierName": "start",
                          "start": Position {
                            "column": 22,
                            "line": 6,
                          },
                        },
                        "name": "start",
                        "start": 87,
                        "type": "Identifier",
                      },
                      "start": 73,
                      "type": "AssignmentExpression",
                    },
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 27,
                        "line": 6,
                      },
                      "filename": undefined,
                      "identifierName": undefined,
                      "start": Position {
                        "column": 8,
                        "line": 6,
                      },
                    },
                    "start": 73,
                    "type": "ExpressionStatement",
                  },
                  Node {
                    "end": 116,
                    "expression": Node {
                      "end": 116,
                      "left": Node {
                        "computed": false,
                        "end": 110,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 17,
                            "line": 7,
                          },
                          "filename": undefined,
                          "identifierName": undefined,
                          "start": Position {
                            "column": 8,
                            "line": 7,
                          },
                        },
                        "object": Node {
                          "end": 105,
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 12,
                              "line": 7,
                            },
                            "filename": undefined,
                            "identifierName": undefined,
                            "start": Position {
                              "column": 8,
                              "line": 7,
                            },
                          },
                          "start": 101,
                          "type": "ThisExpression",
                        },
                        "property": Node {
                          "end": 110,
                          "id": Node {
                            "end": 110,
                            "loc": SourceLocation {
                              "end": Position {
                                "column": 17,
                                "line": 7,
                              },
                              "filename": undefined,
                              "identifierName": "end",
                              "start": Position {
                                "column": 14,
                                "line": 7,
                              },
                            },
                            "name": "end",
                            "start": 107,
                            "type": "Identifier",
                          },
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 17,
                              "line": 7,
                            },
                            "filename": undefined,
                            "identifierName": undefined,
                            "start": Position {
                              "column": 13,
                              "line": 7,
                            },
                          },
                          "start": 106,
                          "type": "PrivateName",
                        },
                        "start": 101,
                        "type": "MemberExpression",
                      },
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 23,
                          "line": 7,
                        },
                        "filename": undefined,
                        "identifierName": undefined,
                        "start": Position {
                          "column": 8,
                          "line": 7,
                        },
                      },
                      "operator": "=",
                      "right": Node {
                        "end": 116,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 23,
                            "line": 7,
                          },
                          "filename": undefined,
                          "identifierName": "end",
                          "start": Position {
                            "column": 20,
                            "line": 7,
                          },
                        },
                        "name": "end",
                        "start": 113,
                        "type": "Identifier",
                      },
                      "start": 101,
                      "type": "AssignmentExpression",
                    },
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 23,
                        "line": 7,
                      },
                      "filename": undefined,
                      "identifierName": undefined,
                      "start": Position {
                        "column": 8,
                        "line": 7,
                      },
                    },
                    "start": 101,
                    "type": "ExpressionStatement",
                  },
                ],
                "directives": Array [],
                "end": 122,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 5,
                    "line": 8,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 28,
                    "line": 5,
                  },
                },
                "start": 63,
                "type": "BlockStatement",
              },
              "computed": false,
              "end": 122,
              "generator": false,
              "id": null,
              "key": Node {
                "end": 50,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 15,
                    "line": 5,
                  },
                  "filename": undefined,
                  "identifierName": "constructor",
                  "start": Position {
                    "column": 4,
                    "line": 5,
                  },
                },
                "name": "constructor",
                "start": 39,
                "type": "Identifier",
              },
              "kind": "constructor",
              "loc": SourceLocation {
                "end": Position {
                  "column": 5,
                  "line": 8,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 4,
                  "line": 5,
                },
              },
              "params": Array [
                Node {
                  "end": 56,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 21,
                      "line": 5,
                    },
                    "filename": undefined,
                    "identifierName": "start",
                    "start": Position {
                      "column": 16,
                      "line": 5,
                    },
                  },
                  "name": "start",
                  "start": 51,
                  "type": "Identifier",
                },
                Node {
                  "end": 61,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 26,
                      "line": 5,
                    },
                    "filename": undefined,
                    "identifierName": "end",
                    "start": Position {
                      "column": 23,
                      "line": 5,
                    },
                  },
                  "name": "end",
                  "start": 58,
                  "type": "Identifier",
                },
              ],
              "start": 39,
              "static": false,
              "type": "ClassMethod",
            },
            Node {
              "async": false,
              "body": Node {
                "body": Array [],
                "directives": Array [],
                "end": 140,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 5,
                    "line": 10,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 10,
                    "line": 9,
                  },
                },
                "start": 133,
                "type": "BlockStatement",
              },
              "computed": false,
              "end": 140,
              "generator": false,
              "id": null,
              "key": Node {
                "end": 131,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 8,
                    "line": 9,
                  },
                  "filename": undefined,
                  "identifierName": "test",
                  "start": Position {
                    "column": 4,
                    "line": 9,
                  },
                },
                "name": "test",
                "start": 127,
                "type": "Identifier",
              },
              "kind": "method",
              "loc": SourceLocation {
                "end": Position {
                  "column": 5,
                  "line": 10,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 4,
                  "line": 9,
                },
              },
              "params": Array [],
              "start": 127,
              "static": false,
              "type": "ClassMethod",
            },
            Node {
              "async": false,
              "body": Node {
                "body": Array [
                  Node {
                    "end": 209,
                    "expression": Node {
                      "arguments": Array [
                        Node {
                          "end": 208,
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 18,
                              "line": 13,
                            },
                            "filename": undefined,
                            "identifierName": "range",
                            "start": Position {
                              "column": 13,
                              "line": 13,
                            },
                          },
                          "name": "range",
                          "start": 203,
                          "type": "Identifier",
                        },
                      ],
                      "callee": Node {
                        "end": 202,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 12,
                            "line": 13,
                          },
                          "filename": undefined,
                          "identifierName": "call",
                          "start": Position {
                            "column": 8,
                            "line": 13,
                          },
                        },
                        "name": "call",
                        "start": 198,
                        "type": "Identifier",
                      },
                      "end": 209,
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 19,
                          "line": 13,
                        },
                        "filename": undefined,
                        "identifierName": undefined,
                        "start": Position {
                          "column": 8,
                          "line": 13,
                        },
                      },
                      "start": 198,
                      "type": "CallExpression",
                    },
                    "leadingComments": Array [
                      Object {
                        "end": 189,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 28,
                            "line": 12,
                          },
                          "filename": undefined,
                          "identifierName": undefined,
                          "start": Position {
                            "column": 8,
                            "line": 12,
                          },
                        },
                        "start": 169,
                        "type": "CommentLine",
                        "value": " this.test(range);",
                      },
                    ],
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 19,
                        "line": 13,
                      },
                      "filename": undefined,
                      "identifierName": undefined,
                      "start": Position {
                        "column": 8,
                        "line": 13,
                      },
                    },
                    "start": 198,
                    "type": "ExpressionStatement",
                  },
                  Node {
                    "alternate": null,
                    "consequent": Node {
                      "argument": Node {
                        "end": 263,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 53,
                            "line": 14,
                          },
                          "filename": undefined,
                          "identifierName": undefined,
                          "start": Position {
                            "column": 48,
                            "line": 14,
                          },
                        },
                        "start": 258,
                        "type": "BooleanLiteral",
                        "value": false,
                      },
                      "end": 263,
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 53,
                          "line": 14,
                        },
                        "filename": undefined,
                        "identifierName": undefined,
                        "start": Position {
                          "column": 41,
                          "line": 14,
                        },
                      },
                      "start": 251,
                      "type": "ReturnStatement",
                    },
                    "end": 263,
                    "loc": SourceLocation {
                      "end": Position {
                        "column": 53,
                        "line": 14,
                      },
                      "filename": undefined,
                      "identifierName": undefined,
                      "start": Position {
                        "column": 8,
                        "line": 14,
                      },
                    },
                    "start": 218,
                    "test": Node {
                      "argument": Node {
                        "arguments": Array [
                          Node {
                            "end": 247,
                            "loc": SourceLocation {
                              "end": Position {
                                "column": 37,
                                "line": 14,
                              },
                              "filename": undefined,
                              "identifierName": "range",
                              "start": Position {
                                "column": 32,
                                "line": 14,
                              },
                            },
                            "name": "range",
                            "start": 242,
                            "type": "Identifier",
                          },
                        ],
                        "callee": Node {
                          "end": 241,
                          "loc": SourceLocation {
                            "end": Position {
                              "column": 31,
                              "line": 14,
                            },
                            "filename": undefined,
                            "identifierName": "hasInstance",
                            "start": Position {
                              "column": 20,
                              "line": 14,
                            },
                          },
                          "name": "hasInstance",
                          "start": 230,
                          "type": "Identifier",
                        },
                        "end": 248,
                        "extra": Object {
                          "parenStart": 223,
                          "parenthesized": true,
                        },
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 38,
                            "line": 14,
                          },
                          "filename": undefined,
                          "identifierName": undefined,
                          "start": Position {
                            "column": 14,
                            "line": 14,
                          },
                        },
                        "start": 224,
                        "type": "CallExpression",
                      },
                      "end": 249,
                      "loc": SourceLocation {
                        "end": Position {
                          "column": 39,
                          "line": 14,
                        },
                        "filename": undefined,
                        "identifierName": undefined,
                        "start": Position {
                          "column": 12,
                          "line": 14,
                        },
                      },
                      "operator": "!",
                      "prefix": true,
                      "start": 222,
                      "type": "UnaryExpression",
                    },
                    "trailingComments": Array [
                      Object {
                        "end": 338,
                        "loc": SourceLocation {
                          "end": Position {
                            "column": 74,
                            "line": 15,
                          },
                          "filename": undefined,
                          "identifierName": undefined,
                          "start": Position {
                            "column": 8,
                            "line": 15,
                          },
                        },
                        "start": 272,
                        "type": "CommentLine",
                        "value": " return this.#start === range.#start && this.#end === range.#end",
                      },
                    ],
                    "type": "IfStatement",
                  },
                ],
                "directives": Array [],
                "end": 344,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 5,
                    "line": 16,
                  },
                  "filename": undefined,
                  "identifierName": undefined,
                  "start": Position {
                    "column": 18,
                    "line": 11,
                  },
                },
                "start": 159,
                "type": "BlockStatement",
              },
              "computed": false,
              "end": 344,
              "generator": false,
              "id": null,
              "key": Node {
                "end": 151,
                "loc": SourceLocation {
                  "end": Position {
                    "column": 10,
                    "line": 11,
                  },
                  "filename": undefined,
                  "identifierName": "equals",
                  "start": Position {
                    "column": 4,
                    "line": 11,
                  },
                },
                "name": "equals",
                "start": 145,
                "type": "Identifier",
              },
              "kind": "method",
              "loc": SourceLocation {
                "end": Position {
                  "column": 5,
                  "line": 16,
                },
                "filename": undefined,
                "identifierName": undefined,
                "start": Position {
                  "column": 4,
                  "line": 11,
                },
              },
              "params": Array [
                Node {
                  "end": 157,
                  "loc": SourceLocation {
                    "end": Position {
                      "column": 16,
                      "line": 11,
                    },
                    "filename": undefined,
                    "identifierName": "range",
                    "start": Position {
                      "column": 11,
                      "line": 11,
                    },
                  },
                  "name": "range",
                  "start": 152,
                  "type": "Identifier",
                },
              ],
              "start": 145,
              "static": false,
              "type": "ClassMethod",
            },
          ],
          "end": 346,
          "loc": SourceLocation {
            "end": Position {
              "column": 1,
              "line": 17,
            },
            "filename": undefined,
            "identifierName": undefined,
            "start": Position {
              "column": 12,
              "line": 2,
            },
          },
          "start": 13,
          "type": "ClassBody",
        },
        "end": 346,
        "id": Node {
          "end": 12,
          "loc": SourceLocation {
            "end": Position {
              "column": 11,
              "line": 2,
            },
            "filename": undefined,
            "identifierName": "Range",
            "start": Position {
              "column": 6,
              "line": 2,
            },
          },
          "name": "Range",
          "start": 7,
          "type": "Identifier",
        },
        "loc": SourceLocation {
          "end": Position {
            "column": 1,
            "line": 17,
          },
          "filename": undefined,
          "identifierName": undefined,
          "start": Position {
            "column": 0,
            "line": 2,
          },
        },
        "start": 1,
        "superClass": null,
        "type": "ClassDeclaration",
      },
    ],
    "directives": Array [],
    "end": 348,
    "interpreter": null,
    "loc": SourceLocation {
      "end": Position {
        "column": 0,
        "line": 19,
      },
      "filename": undefined,
      "identifierName": undefined,
      "start": Position {
        "column": 0,
        "line": 1,
      },
    },
    "sourceType": "module",
    "start": 0,
    "type": "Program",
  },
  "start": 0,
  "type": "File",
}
`;
